37.md

```diff
@@ -0,0 +1,95 @@
+NIP-37
+======
+
+Methods for dealing with lost or compromised keys
+-------------------------------------------------
+
+`draft` `optional` `author:fiatjaf`
+
+This NIP defines a series of methods that can be used to make catastrophic key loss events less horrible.
+
+The intention here is not to claim that there is a generalized "key rotation" scheme on Nostr, but still make it
+so that in most cases people are able to move to new keys and not lost 100% of their followers and reputation
+when that happens, and also that it is possible to prevent most of the reputational damage that could be
+caused by having an evil hacker take control of someone else's keys.
+
+The implementation of this NIP could be done directly in social clients or in dedicated clients that users would
+visit from time to time in order to have an automated scan be done on their contacts and then the dedicated app
+could suggest migrating from old keys to new keys, or suggest stop following compromised keys, or even perform
+these migrations automatically.
+
+### Method A: Simple Key Deletion
+
+Based on https://github.com/nostr-protocol/nips/pull/377, the idea is that someone, after noticing that their key
+has been compromised and deciding that it is not possible to use any of the other methods here described, can
+proceed to publish the following event:
+
+```js
+{
+  "kind": 10529,
+  "pubkey": "<compromised-pubkey>",
+  "tags": [
+    ["reason", "I typed my key on anigma"]
+
+    // this is unnecessary, but it's an extra requirement to prevent people from making a 10529 event by chance
+    ["key-compromised"],
```

Comment:
I think this is can be implicit in the value of the “kind” itself and rather make it mandatory for the client to ask a “challenge” to sign a kind 10529 with the compromised pubkey. E.g a “challenge” can be the example given after of requiring a user to type a message. Other “challenge” could be a pre-registered password in the client hard state or really any existent password protocol (e.g a time-based one-time based password).

See also PGP revocation format, it’s just a dummy 0x20 signature type to announce revocation: https://www.rfc-editor.org/rfc/rfc4880#section-5.2.1

Reply:
I am just worried because I use this https://github.com/fiatjaf/nak CLI tool to generate events and it's harmless 99.9% of the times, but if me or someone else typo or type at random the deletion kind they are gone from the Earth.

Reply:
Yeah but it's easier to accidentally copy and paste your private key into a public chatroom. Then it's also gone forever with even less effort.

Reply:
So you're saying we should have this NIP specify that you should type your private key in a public chatroom?

Reply:
> I am just worried because I use this https://github.com/fiatjaf/nak CLI tool to generate events and it's harmless 99.9% of the times, 

Yeah I would expect with time that type of tools to request a challenge like totp: https://www.rfc-editor.org/rfc/rfc6238 (google authenticator based on that). For sure key compromise event generation gonna stay a worrying operation for user.